// react-jinke-music-player/lib/components/PlayerMobile.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

var _PlayModeTip = _interopRequireDefault(require("./PlayModeTip"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var prefix = 'react-jinke-music-player-mobile';

var PlayerMobile = function PlayerMobile(_ref) {
  var name = _ref.name,
      cover = _ref.cover,
      singer = _ref.singer,
      playing = _ref.playing,
      duration = _ref.duration,
      currentTime = _ref.currentTime,
      loading = _ref.loading,
      themeSwitch = _ref.themeSwitch,
      progressBar = _ref.progressBar,
      openAudioListsPanel = _ref.openAudioListsPanel,
      audioPrevPlay = _ref.audioPrevPlay,
      audioNextPlay = _ref.audioNextPlay,
      playMode = _ref.playMode,
      onClose = _ref.onClose,
      playModeTipVisible = _ref.playModeTipVisible,
      currentPlayModeName = _ref.currentPlayModeName,
      extendsContent = _ref.extendsContent,
      onPlay = _ref.onPlay,
      glassBg = _ref.glassBg,
      onCoverClick = _ref.onCoverClick,
      autoHiddenCover = _ref.autoHiddenCover,
      icon = _ref.icon,
      locale = _ref.locale,
      toggleMode = _ref.toggleMode,
      renderAudioTitle = _ref.renderAudioTitle,
      shouldShowPlayIcon = _ref.shouldShowPlayIcon,
      isResetCoverRotate = _ref.isResetCoverRotate;
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: (0, _classnames["default"])(prefix, {
      'default-bg': !glassBg,
      'glass-bg': glassBg
    })
  }, /*#__PURE__*/_react["default"].createElement(_PlayModeTip["default"], {
    prefix: prefix,
    visible: playModeTipVisible,
    title: playMode,
    text: currentPlayModeName
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-header group")
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-header-title"),
    title: name
  }, renderAudioTitle()), toggleMode && /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-header-right"),
    onClick: onClose
  }, icon.close)), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-singer text-center group")
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "".concat(prefix, "-singer-name"),
    title: singer
  }, singer)), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-switch text-center group")
  }, themeSwitch), (!autoHiddenCover || autoHiddenCover && cover) && /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-cover text-center"),
    onClick: function onClick() {
      return onCoverClick();
    }
  }, /*#__PURE__*/_react["default"].createElement("img", {
    src: cover,
    alt: "cover",
    className: (0, _classnames["default"])('cover', {
      'img-rotate-pause': !playing || !cover,
      'img-rotate-reset': isResetCoverRotate
    })
  })), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-progress group")
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "current-time"
  }, loading ? '--' : currentTime), /*#__PURE__*/_react["default"].createElement("span", {
    className: "".concat(prefix, "-progress-bar")
  }, progressBar), /*#__PURE__*/_react["default"].createElement("span", {
    className: "duration text-right"
  }, loading ? '--' : duration)), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-toggle text-center group")
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "group prev-audio",
    title: locale.previousTrackText,
    onClick: audioPrevPlay
  }, icon.prev), loading ? /*#__PURE__*/_react["default"].createElement("span", {
    className: "group loading-icon"
  }, icon.loading) : /*#__PURE__*/_react["default"].createElement("span", {
    className: "group play-btn",
    title: shouldShowPlayIcon ? locale.clickToPlayText : locale.clickToPauseText,
    onClick: onPlay
  }, shouldShowPlayIcon ? icon.play : icon.pause), /*#__PURE__*/_react["default"].createElement("span", {
    className: "group next-audio",
    title: locale.nextTrackText,
    onClick: audioNextPlay
  }, icon.next)), /*#__PURE__*/_react["default"].createElement("div", {
    className: "".concat(prefix, "-operation group")
  }, /*#__PURE__*/_react["default"].createElement("ul", {
    className: "items"
  }, [playMode, icon.download, icon.reload, icon.lyric].filter(_react.isValidElement).map(function (item) {
    return /*#__PURE__*/_react["default"].createElement("li", {
      className: "item",
      key: item.props.className
    }, item);
  }), extendsContent, /*#__PURE__*/_react["default"].createElement("li", {
    className: "item",
    onClick: openAudioListsPanel
  }, icon.playLists))));
};

PlayerMobile.defaultProps = {
  icon: {},
  renderAudioTitle: function renderAudioTitle() {}
};

var _default = /*#__PURE__*/(0, _react.memo)(PlayerMobile);

exports["default"] = _default;
